### DISCUSSION TITLE
How hard is the 4GB free requirement?

### DISCUSSION THREAD

**Discussion by llucax** [nextcloud/recognize#732] [Q&A] [answered]
So the readme says this requires 4GB of free memory. I understand by that that a system with 4GB won't be enough. What happens if I install in a system with leaa memory than that (but has swap)? Will this run every time I upload a supporter file?

I have a personal instance in a server with 2GB, where only a few photos and videos per day are uploaded. Would I be able to use it if it swaps or would it be completely unusable?

Thanks!

**Comment by marcelklehr** [comment, marked as answer]
> I understand by that that a system with 4GB won't be enough.

As far as I remember, the 4GB requirement includes some wiggle room. So, if you have swap / nothing else running, 4GB might be workable. I'm sceptical about 2GB, though. In the end I suggest you try it out and report back here :)

**Reply by llucax** [reply]
Well, I tried with the 1 GB and it works. it's very slow but it works. it takes I would say about 10, 15 minutes to process one image, maybe (I'm guessing, I couldn't find any logs where i can see more precisely), with a lot of swapping. (On an Intel Xeon E5 shared vCPU in a VPS)

I'm not sure what would be the normal time when there is no swapping to process one image.

Also I don't know much about ML but I thought many face and other image recognition models, once trained can be super fast to classify the images and don't require a lot of resources. I just have a misconception or is there anything else I'm misunderstanding?

Also, is there a (easy) way to offload this ML processing to a different server than the one running nextcloud itself?

Thanks!

**Reply by llucax** [reply]
My previous test was enabling all types or recognition. Using only face and video recognition with 2251 files (from which only a few dozens are videos), after importing the files manually and running `files:scan` and then `recognize:classify`, it took about 3h to do the classification. That is about 5s per image, and even when there was some swapping in this case just with 1GB, it wasn't much, so I'm pretty OK with the results.

The object recognition seems to be much more memory hungry and slower. Does this make sense? I guess judging from the recommended values for the batches, it sort of does, as face recognition can handle 5x more than the other ones. BTW, my server uses native TensorFlow operation and since I had only 1GB I configured it to batch 1/10 of the recommended WASM value, just to be on the safe side (so 5 for face, and 2 for the rest). Although when running `recognize:classify` I guess those values are not used, right?
